src/arithmetic/rewrite_simplify.cc

```diff
@@ -0,0 +1,646 @@
+
+    TVM_TRY_REWRITE_IF(min(b1, b2) - min(s1, s2), b1 - s2,
+                       CanProveEqual(((b1 - s2) - (b2 - s1)).Eval(), 0));
+    TVM_TRY_REWRITE_IF(max(b1, b2) - max(s1, s2), b1 - s1,
+                       CanProveEqual(((b1 - s1) - (b2 - s2)).Eval(), 0));
+    TVM_TRY_REWRITE_IF(max(b1, b2) - max(s1, s2), b1 - s2,
+                       CanProveEqual(((b1 - s2) - (b2 - s1)).Eval(), 0));
+
+    // modular-div simplification
+    // Always pre-condition on positive integer domain
+    TVM_TRY_REWRITE_IF(x - (x / c1) * c1, x % c1,
+                       CanProveGreaterEqual(x.Eval(), 0) && c1.Eval()->value > 0);
+    TVM_TRY_REWRITE_IF((x / c1) * c1 - x, 0 - (x % c1),
+                       CanProveGreaterEqual(x.Eval(), 0) && c1.Eval()->value > 0);
+    TVM_TRY_REWRITE_IF((x + c1) / c3  - (x + c2) / c3,
+                       ((x + (c1 % c3)) % c3 + (c1 - c2)) / c3,
+                       CanProveGreaterEqual(x.Eval(), -c2.Eval()->value) &&
+                       c1.Eval()->value >= c2.Eval()->value &&
+                       c3.Eval()->value > 0);
+    TVM_TRY_REWRITE_IF((x + c1) / c3  - x / c3,
+                       ((x + (c1 % c3)) % c3 + c1) / c3,
+                       CanProveGreaterEqual(x.Eval(), 0) &&
+                       c1.Eval()->value >= 0 &&
+                       c3.Eval()->value > 0);
+    // canonicalization rule
+    // will try rewrite again after canonicalization.
+    TVM_TRY_REWRITE(x - c1, x + (0 - c1));
+    TVM_TRY_RECURSIVE_REWRITE((x + c1) - y, (x - y) + c1);
+    TVM_TRY_RECURSIVE_REWRITE(x - (y - z), (x + z) - y);
+    TVM_TRY_RECURSIVE_REWRITE(x - y * c1, x + y * (0 - c1));
+  }
+
+  // condition rules.
+  TVM_TRY_REWRITE(select(x, b1, b2) - select(x, s1, s2),
+                  select(x, b1 - s1, b2 - s2));
+  TVM_TRY_REWRITE(select(x, y, z) - z,
+                  select(x, y - z, ZeroWithTypeLike(z)));
+  TVM_TRY_REWRITE(select(x, y, z) - y,
+                  select(x, ZeroWithTypeLike(y), z - y));
+  return ret;
+}
+
+Expr RewriteSimplifier::Impl::
+Mutate_(const Mul* op, const Expr& self) {
+  Expr ret = IRMutator::Mutate_(op, self);
+  op = ret.as<Mul>();
+  Expr const_res = TryConstFold<Mul>(op->a, op->b);
+  if (const_res.defined()) return const_res;
+  // Pattern var to match any expression
+  PVar<Expr> x, y, z, b1, b2, s1, s2;
+  // Pattern var match IntImm
+  PVar<Integer> c1, c2;
+  // Pattern var for lanes in broadcast and ramp
+  PVar<int> lanes;
+  // Vector rules
+  if (op->type.lanes() != 1) {
+    TVM_TRY_REWRITE(broadcast(x, lanes) * broadcast(y, lanes),
+                    broadcast(x * y, lanes));
+    TVM_TRY_REWRITE(ramp(b1, s1, lanes) * broadcast(x, lanes),
+                    ramp(b1 * x, s1 * x, lanes));
+    TVM_TRY_REWRITE(broadcast(x, lanes) * ramp(b1, s1, lanes),
+                    ramp(b1 * x, s1 * x, lanes));
+  }
+
+  if (IsIndexType(op->type)) {
+    // constant simplification rule
+    TVM_TRY_REWRITE((x + c1) * c2, x * c2 + c1 * c2);
```

Comment:
Do this rule contradicts with `TVM_TRY_REWRITE(x * y + x * z, x * (y + z))` from the `Add` section? Does RewriteSimplifier stabilizes if repeatedly applied many times?

Reply:
This rewrite only works for constant and note that c1 * c2 will get automatically constant folded